geom/alg_convex_hull.go

```diff
@@ -48,108 +48,94 @@ func convexHull(g Geometry) Geometry {
 		return g
 	}
 	pts := g.convexHullPointSet()
-	hull := grahamScan(pts)
-	switch len(hull) {
+	// TODO: length may not be a good signal to indicate
+	// how to convert XY to Geometry. Need to revisit
+	switch len(pts) {
 	case 0:
 		return NewGeometryCollection(nil)
 	case 1:
-		return NewPointXY(hull[0])
+		return NewPointXY(pts[0])
 	case 2:
+		if pts[0].Equals(pts[1]) {
+			return NewPointXY(pts[0])
+		}
 		ln, err := NewLineC(
-			Coordinates{hull[0]},
-			Coordinates{hull[1]},
+			Coordinates{pts[0]},
+			Coordinates{pts[1]},
 		)
 		if err != nil {
-			panic("bug in grahamScan routine - output 2 coincident points")
+			panic(err)
 		}
 		return ln
-	default:
-		coords := [][]Coordinates{make([]Coordinates, len(hull))}
-		for i := range hull {
-			coords[0][i] = Coordinates{XY: hull[i]}
-		}
-		poly, err := NewPolygonC(coords)
-		if err != nil {
-			panic(fmt.Errorf("bug in grahamScan routine - didn't produce a valid polygon: %v", err))
-		}
-		return poly
 	}
-}
-
-type pointStack []XY
-
-func (s *pointStack) push(p XY) {
-	(*s) = append(*s, p)
-}
 
-func (s *pointStack) pop() XY {
-	p := s.top()
-	(*s) = (*s)[:len(*s)-1]
-	return p
-}
+	// TODO: check if the points are all the same
+	isSame := true
+	for i := 1; i < len(pts); i++ {
+		if !pts[0].Equals(pts[i]) {
+			isSame = false
+			break
+		}
+	}
+	if isSame {
+		return NewPointXY(pts[0])
+	}
 
-func (s *pointStack) top() XY {
-	return (*s)[len(*s)-1]
-}
+	cl, ok := collinearLine(pts)
+	if ok {
+		return cl
+	}
 
-func (s *pointStack) underTop() XY {
-	return (*s)[len(*s)-2]
+	hull := grahamScan(pts)
+	coords := [][]Coordinates{make([]Coordinates, len(hull))}
+	for i := range hull {
+		coords[0][i] = Coordinates{XY: hull[i]}
+	}
+	poly, err := NewPolygonC(coords)
+	if err != nil {
+		panic(fmt.Errorf("bug in grahamScan routine - didn't produce a valid polygon: %v", err))
+	}
+	return poly
 }
 
 // grahamScan returns the convex hull of the input points. It will either
 // represent the empty set (zero points), a point (one point), a line (2
 // points), or a closed polygon (>= 3 points).
 func grahamScan(ps []XY) []XY {
-	if len(ps) <= 1 {
-		return ps
-	}
 
 	sortByPolarAngle(ps)
 
 	// Append the lowest-then-leftmost point so that the polygon will be closed.
-	ps = append(ps, ps[0])
-
-	// Populate the stack with the first 2 distict points.
-	var i int // tracks progress through the ps slice
-	var stack pointStack
-	stack.push(ps[0])
-	i++
-	for i < len(ps) && len(stack) < 2 {
-		if !stack.top().Equals(ps[i]) {
-			stack.push(ps[i])
-		}
-		i++
-	}
-
-	for i < len(ps) {
-		ori := orientation(stack.underTop(), stack.top(), ps[i])
+	resultStack := make([]XY, 0, len(ps))
```

Comment:
What are your thoughts on adding back in the `pointStack` abstraction? I think having the `pop`, `push`, `top` and `underTop` abstractions really helps to make the algorithm more readable.